File: docs/algorithms/code/algorithms/findBottomLeftValue.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

// 描述：需要找到最后一层最左边的节点的值
// 思路：层序遍历，找到最后一层的第一个节点的值

function findBottomLeftValue(root: TreeNode | null): number {
  const stack: TreeNode[] = [root];

  let res = 0;

  while (stack.length) {
    const len = stack.length;
    for (let i = 0; i < len; i++) {
      const node = stack.shift();
      if (i === 0) {
        res = node.val;
      }
      if (node.left) {
        stack.push(node.left);
      }
      if (node.right) {
        stack.push(node.right);
      }
    }
  }

  return res;
};

File: docs/algorithms/code/algorithms/findBottomLeftValue.go
// 思路：层序遍历，最后一层第一个节点
// 复杂度：时间复杂度 O(n)，空间复杂度 O(n)
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func findBottomLeftValue(root *TreeNode) int {
	if root == nil {
		return 0
	}

	stack := []*TreeNode{root}
	res := 0;
	for len(stack) > 0 {
		size := len(stack)
		res = stack[0].Val
		for i := 0; i < size; i++ {
			node := stack[0]
			stack = stack[1:]
			if node.Left != nil {
				stack = append(stack, node.Left)
			}
			if node.Right != nil {
				stack = append(stack, node.Right)
			}
		}
	}
	return res
}